Installation Instructions (#29)
emersion/soju: INSTALL.md

# Requirements

## Go runtime

Follow the instructions for installing go found at <https://golang.org/doc/install>

Running `go version` will confirm go has been installed successfully and you can continue.

## Install build tools

> Soju uses a Makefile for building the project and needs gcc to build the go-sqlite3 library. On Debian based systems, including Ubuntu, these can be installed by running `sudo apt install make gcc`

Reviewer: ```suggestion
Soju uses a Makefile for building the project and needs an C89 compiler to build the go-sqlite3 library (e.g. gcc). On Debian based systems, including Ubuntu, these can be installed by running `sudo apt install make gcc`
```

## scdoc (optional)

Soju builds man pages using the scdoc command which can be installed with `sudo apt install scdoc` This is optional.

# Clone the soju repository

`git clone https://git.sr.ht/~emersion/soju` will create a soju directory and clone the project into it

# Building and Installing

## Run `make`

Enter the soju directory and issue the `make` command.  The included Makefile will perform the necessary steps to build the project. When make finishes you will have `soju` and `sojuctl` executables plus man pages for each.

## `make install`

Run `sudo make install` to install soju system wide.

1.  The executables will be placed in /usr/local/bin
2.  A default config file will be located at /etc/soju/config
3.  A data directory will be created at /var/lib/soju
4.  Man pages will be copied to /usr/local/share/man

You do not need to install soju system wide, but the remainder of this document assumes you have, so modify your configuration files accordingly.

> # Service Confugration

Reviewer 2: [typo] Confugration → Configuration

## Add a system user

`sudo useradd -r soju` will create a system user account named soju. The user will not have a password set and cannot be used to log in to the system.

## Grant soju user ownership of /var/lib/soju

`sudo chown soju /var/lib/soju`

## Generate certificates

Using openssl generate a self signed certifcate clients will use to confirm they're connecting to your soju instance.

`sudo -u soju openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout /var/lib/soju/key.pem -out /var/lib/soju/cert.pem`

The command will generate certificates, owned by our soju user, and write them to the soju data directory.

> ## Confgiure Soju

Reviewer 2: [typo] Confgiure → Configure

Edit the configuration file located at `/etc/soju/config` adding the TLS directive.

```
db sqlite3 /var/lib/soju/main.db 
log fs /var/lib/soju/logs 
tls /var/lib/soju/cert.pem /var/lib/soju/key.pem 
```

> By default soju will listen on all interfaces. If you want it to listen only on a specific ip address add the line `listen x.y.z.w:6697`.  You can optionally specify the hostname soju should use with `hostname example.org`.

Reviewer: Could you also add a ref here

# Building and Installing

## Run `make`

Enter the soju directory and issue the `make` command.  The included Makefile will perform the necessary steps to build the project. When make finishes you will have `soju` and `sojuctl` executables plus man pages for each.

## `make install`

Run `sudo make install` to install soju system wide.

1.  The executables will be placed in /usr/local/bin
2.  A default config file will be located at /etc/soju/config
3.  A data directory will be created at /var/lib/soju
4.  Man pages will be copied to /usr/local/share/man

You do not need to install soju system wide, but the remainder of this document assumes you have, so modify your configuration files accordingly.

# Service Confugration

## Add a system user

`sudo useradd -r soju` will create a system user account named soju. The user will not have a password set and cannot be used to log in to the system.

## Grant soju user ownership of /var/lib/soju

`sudo chown soju /var/lib/soju`

## Generate certificates

Using openssl generate a self signed certifcate clients will use to confirm they're connecting to your soju instance.

> `sudo -u soju openssl req -newkey rsa:2048 -new -nodes -x509 -days \\ 3650 -keyout /var/lib/soju/key.pem -out /var/lib/soju/cert.pem`

Reviewer: ```suggestion
`sudo -u soju openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout /var/lib/soju/key.pem -out /var/lib/soju/cert.pem`
```

The command will generate certificates, owned by our soju user, and write them to the soju data directory.

## Confgiure Soju

> Edit the configuration file located at /etc/soju/config

Reviewer: ```suggestion
Edit the configuration file located at `/etc/soju/config`
```

```
db sqlite3 /var/lib/soju/main.db 
log fs /var/lib/soju/logs 
tls /var/lib/soju/cert.pem /var/lib/soju/key.pem 

> hostname ****example.org**** 

Reviewer: the `hostname` directive isn't needed and could be removed from this guide for clarity?

```

> By default soju will listen on all interfaces. If you want it to listen only on a specific ip address add the linke `listen x.y.z.w:6697` to the config file where x.y.z.w is the ip address of that interface.

Reviewer: ```suggestion
By default soju will listen on all interfaces. If you want it to listen only on a specific ip address add the linke `listen x.y.z.w:6697` to the config file where `x.y.z.w` is the ip address of that interface.
```

## Create systemd service file

A systemd service file will allow systemd to manage the soju daemon service. It will be started on boot, restarted on crash (giving up after a few in a row), and let you restart the service via `systemctl restart soju` Logs from the soju service can be read with `journalctl -u soju`

> 1.  Place the below configuration in the file /etc/systemd/system/soju.service

Reviewer: ```suggestion
1.  Place the below configuration in the file `/etc/systemd/system/soju.service`
```

2.  Execute `sudo systemctl enable soju.service` which will start soju when the system boots
3.  `sudo systemctl start soju.service` will immediately start the service.

```
[Unit] 
Description=Soju bouncer daemon 
After=network.target

[Service] 
Type=simple 
Restart=always 
RestartSec=15 
User=soju 

> ExecStart=/usr/local/bin/soju -config /etc/soju/config

Reviewer: You can also add this line, which will allow cert reload via `systemctl reload soju` (so that you don't have to restart it):

    ExecReload=kill -HUP $MAINPID

See https://git.sr.ht/~emersion/soju/tree/master/item/doc/soju.1.scd#L47

[Install] 
WantedBy=multi-user.target
```

# Using Soju

## Add your first user

> `sudo -u soju sojuctl -config /etc/soju/config create-user ****myuser**** \\ -admin`

Reviewer: ```suggestion
`sudo -u soju sojuctl -config /etc/soju/config create-user ****myuser**** -admin`
```

> You will be prompted for your user password. You will use this username and password to connect to your soju service via your IRC client. The -admin flag creates an admin user.

Reviewer: ```suggestion
You will be prompted for your user password. You will use this username and password to connect to your soju service via your IRC client. The `-admin` flag creates an admin user.
```

## Connect to the running soju bouncer

> Connect with the username and password you've created in the previous step. Make sure your clienet knows to connect on port 6697 and that ssl/tsl is enabled for the connection. You will be prompted to accept the self-signed certifcate you created earlier. This will identify your soju instance to your client.

Reviewer: ```suggestion
Connect with the username and password you've created in the previous step. Make sure your client knows to connect on port 6697 and that TLS is enabled for the connection. You will be prompted to accept the self-signed certificate you created earlier. This will identify your soju instance to your client.
```

## Adding a Server

Your soju instance can be managed by messaging the bot bouncerserv. `/msg bouncerserv help` for a list of available commands.

To add a new server, Libera.chat for example, use the command `/msg bouncerserv newtwork create -addr irc.libera.chat -name libera`

> ## Set your sasl credentials

Reviewer: ```suggestion
## Set your SASL credentials
```

`/msg bouncerserv sasl set-plain libera user pass`

> Notice when creating the network we named it libera, so we can use the shorter name when setting sasl commands.

Reviewer: ```suggestion
Notice when creating the network we named it libera, so we can use the shorter name when setting SASL commands.
```

> ## Single Upstream vs Multi Upstream Mode

Reviewer: Can you put/merge this and the following sections in `doc/getting-started.md` instead?
